## Project: Milestone 2

### DSC 540

#### Taniya Adhikari 1/24/2021

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import warnings; warnings.simplefilter('ignore')

### 1. Reading the data

In [2]:
# read the data
df = pd.read_csv("robinhood.csv")
df.head()

simple_name                                name symbol  \
0  Aurora Cannabis  Aurora Cannabis Inc. Common Shares    ACB   
1             Ford                  Ford Motor Company      F   
2               GE            General Electric Company     GE   
3        Microsoft  Microsoft Corporation Common Stock   MSFT   
4            GoPro    GoPro, Inc. Class A Common Stock   GPRO   

     bloomberg_unique  margin_initial_ratio  maintenance_ratio country  \
0  EQ0000000036343385                   1.0               1.00      CA   
1  EQ0010060200001000                   0.5               0.25      US   
2  EQ0010063200001000                   0.5               0.25      US   
3  EQ0010174300001000                   0.5               0.25      US   
4  EQ0000000017107861                   0.6               0.50      US   

   day_trade_ratio  list_date   type  ...  pe_ratio  shares_outstanding  \
0             0.25  7/10/2014  stock  ...       NaN        1.168980e+09   
1             0.25   1/2/1990  stock  ...  615.3850        3.965050e+09   
2             0.25   1/2/1990  stock  ...  147.6300        8.738434e+09   
3             0.25  9/17/1987  stock  ...   30.0787        7.606050e+09   
4             0.25  6/26/2014  stock  ...       NaN        1.559960e+08   

                                         description  \
0  Aurora Cannabis, Inc. engages in the productio...   
1  Ford Motor Co. engages in the manufacture, dis...   
2  General Electric Co. is a technology and finan...   
3  Microsoft Corp. engages in the development and...   
4  GoPro, Inc. engages in manufacturing and selli...   

                          ceo  headquarters_city  headquarters_state  \
0              Michael Singer           Edmonton             Alberta   
1            James P. Hackett           Dearborn            Michigan   
2  H. Lawrence Culp, Jr., MBA             Boston       Massachusetts   
3               Satya Nadella            Redmond          Washington   
4            Nicholas Woodman          San Mateo          California   

                  sector                          industry  num_employees  \
0     Process Industries  Agricultural Commodities/Milling         2779.0   
1      Consumer Durables                    Motor Vehicles       190000.0   
2  Electronic Technology             Aerospace And Defense       205000.0   
3    Technology Services                 Packaged Software       144000.0   
4      Consumer Durables            Electronics/Appliances          926.0   

   year_founded  
0        2006.0  
1        1903.0  
2        1878.0  
3        1975.0  
4        2002.0  

[5 rows x 33 columns]

In [3]:
# checking data type
def Datatype(df):
    # shape and data types of the data
    print("There are {} rows and {} columns".format(df.shape[0], df.shape[1]))
    print(df.dtypes)

    # select numeric columns
    df_numeric = df.select_dtypes(include=[np.number])
    numeric_cols = df_numeric.columns.values
    print(numeric_cols)

    # select non numeric columns
    df_non_numeric = df.select_dtypes(exclude=[np.number])
    non_numeric_cols = df_non_numeric.columns.values
    print(non_numeric_cols)

In [4]:
Datatype(df)

There are 4610 rows and 33 columns
simple_name                 object
name                        object
symbol                      object
bloomberg_unique            object
margin_initial_ratio       float64
maintenance_ratio          float64
country                     object
day_trade_ratio            float64
list_date                   object
type                        object
default_collar_fraction    float64
open                       float64
high                       float64
low                        float64
volume                       int64
average_volume_2_weeks     float64
average_volume             float64
high_52_weeks              float64
dividend_yield             float64
float                      float64
low_52_weeks               float64
market_cap                 float64
pb_ratio                   float64
pe_ratio                   float64
shares_outstanding         float64
description                 object
ceo                         object
headquarters_city   

### 2. Renaming data headers

In [5]:
# iterating the columns 
for col in df.columns: 
    print(col)

simple_name
name
symbol
bloomberg_unique
margin_initial_ratio
maintenance_ratio
country
day_trade_ratio
list_date
type
default_collar_fraction
open
high
low
volume
average_volume_2_weeks
average_volume
high_52_weeks
dividend_yield
float
low_52_weeks
market_cap
pb_ratio
pe_ratio
shares_outstanding
description
ceo
headquarters_city
headquarters_state
sector
industry
num_employees
year_founded


In [6]:
# renaming headers
df = df.rename(columns={"dividend_yield": "dividend_yield_old", "pb_ratio": "pb_ratio_old","pe_ratio": "pe_ratio_old", "symbol": "ticker", "market_cap": "market_cap_old", "simple_name": "name_short", "name": "name_long", "type": "investment_type", "float": "comp_float"})
df.head()

name_short                           name_long ticker  \
0  Aurora Cannabis  Aurora Cannabis Inc. Common Shares    ACB   
1             Ford                  Ford Motor Company      F   
2               GE            General Electric Company     GE   
3        Microsoft  Microsoft Corporation Common Stock   MSFT   
4            GoPro    GoPro, Inc. Class A Common Stock   GPRO   

     bloomberg_unique  margin_initial_ratio  maintenance_ratio country  \
0  EQ0000000036343385                   1.0               1.00      CA   
1  EQ0010060200001000                   0.5               0.25      US   
2  EQ0010063200001000                   0.5               0.25      US   
3  EQ0010174300001000                   0.5               0.25      US   
4  EQ0000000017107861                   0.6               0.50      US   

   day_trade_ratio  list_date investment_type  ...  pe_ratio_old  \
0             0.25  7/10/2014           stock  ...           NaN   
1             0.25   1/2/1990           stock  ...      615.3850   
2             0.25   1/2/1990           stock  ...      147.6300   
3             0.25  9/17/1987           stock  ...       30.0787   
4             0.25  6/26/2014           stock  ...           NaN   

   shares_outstanding                                        description  \
0        1.168980e+09  Aurora Cannabis, Inc. engages in the productio...   
1        3.965050e+09  Ford Motor Co. engages in the manufacture, dis...   
2        8.738434e+09  General Electric Co. is a technology and finan...   
3        7.606050e+09  Microsoft Corp. engages in the development and...   
4        1.559960e+08  GoPro, Inc. engages in manufacturing and selli...   

                          ceo  headquarters_city  headquarters_state  \
0              Michael Singer           Edmonton             Alberta   
1            James P. Hackett           Dearborn            Michigan   
2  H. Lawrence Culp, Jr., MBA             Boston       Massachusetts   
3               Satya Nadella            Redmond          Washington   
4            Nicholas Woodman          San Mateo          California   

                  sector                          industry  num_employees  \
0     Process Industries  Agricultural Commodities/Milling         2779.0   
1      Consumer Durables                    Motor Vehicles       190000.0   
2  Electronic Technology             Aerospace And Defense       205000.0   
3    Technology Services                 Packaged Software       144000.0   
4      Consumer Durables            Electronics/Appliances          926.0   

   year_founded  
0        2006.0  
1        1903.0  
2        1878.0  
3        1975.0  
4        2002.0  

[5 rows x 33 columns]

### 3. Check for duplicates

In [7]:
print("Ticker Symbol duplicated: {}".format(any(df.ticker.duplicated())))
print("Bloomberg Unique duplicated: {}".format(any(df.bloomberg_unique.duplicated())))

Ticker Symbol duplicated: False
Bloomberg Unique duplicated: False


### 4. Subsetting or Filtering dataset

In [8]:
# subsetting data for only companies with market cap over 1B, only for stocks and necessary columns
def subsetdata(df, col_names, mkCap, stType):
    
    # selecting colums by col_names
    df1 = df[col_names]
    
    # filtering for companies with 1 Billion market Cap
    df2 = df1[(df1[mkCap] >=1e9) & (df1[stType] =="stock")]    
    
    index = df2.index
    number_of_rows = len(index)
    print("total_rows: {}". format(number_of_rows))
            
    return df2

In [9]:
# dropping variables bloomberg_unique ID, open, high, low, volume...and other variables
col_names = ['ticker', 'name_long', 'name_short', 'country', 'margin_initial_ratio', 
             'maintenance_ratio', 'day_trade_ratio', 'list_date', 'investment_type', 
             'default_collar_fraction','dividend_yield_old', 'comp_float', 'market_cap_old', 
             'pb_ratio_old', 'pe_ratio_old', 'shares_outstanding', 'description', 'ceo', 'sector', 
             'industry', 'num_employees', 'year_founded']
mkCap = 'market_cap_old'
stType = 'investment_type'
subset_df = subsetdata(df, col_names, mkCap, stType)
subset_df.shape

total_rows: 1451


(1451, 22)

In [10]:
subset_df.head(2)

ticker                           name_long       name_short country  \
0    ACB  Aurora Cannabis Inc. Common Shares  Aurora Cannabis      CA   
1      F                  Ford Motor Company             Ford      US   

   margin_initial_ratio  maintenance_ratio  day_trade_ratio  list_date  \
0                   1.0               1.00             0.25  7/10/2014   
1                   0.5               0.25             0.25   1/2/1990   

  investment_type  default_collar_fraction  ...  market_cap_old  pb_ratio_old  \
0           stock                     0.05  ...    1.576224e+09      0.536001   
1           stock                     0.05  ...    2.759671e+10      0.883266   

   pe_ratio_old  shares_outstanding  \
0           NaN        1.168980e+09   
1       615.385        3.965050e+09   

                                         description               ceo  \
0  Aurora Cannabis, Inc. engages in the productio...    Michael Singer   
1  Ford Motor Co. engages in the manufacture, dis...  James P. Hackett   

               sector                          industry num_employees  \
0  Process Industries  Agricultural Commodities/Milling        2779.0   
1   Consumer Durables                    Motor Vehicles      190000.0   

  year_founded  
0       2006.0  
1       1903.0  

[2 rows x 22 columns]

Dropped following variables: 
* bloomberg_unique  
* open 
* high 
* low 
* volume 
* average_volume_2_weeks 
* average_volume 
* high_52_weeks 
* low_52_weeks 
* headquarters_city
* headquarters_state

Open, high, low, volume changes daily and since this dataset is at least 8 months old, I will not need these values. I will be fetching these values from th API source. 

### 5. Handling Missing Values

I created a function to check the percentage of missing values in each feature. deleing null values is always risky, it needs careful analysis. I attempt to check which feature has highest missing values an ddepending on it's important I will either delete or keep the features

In [11]:
# check for missing value
def percentMissing(df):
    
    df_numeric = df.select_dtypes(include=[np.number])
    numeric_cols = df_numeric.columns.values
    print(numeric_cols)
    
    # % of missing data
    for col in df.columns:
        # create missing indicator for features with missing data
        missing = df[col].isnull()
        pct_missing = np.mean(missing)*100
        print('{} - {}%'.format(col, round(pct_missing)))
        num_missing = np.sum(missing)
        
    return df

In [12]:
df_missing = percentMissing(subset_df)

['margin_initial_ratio' 'maintenance_ratio' 'day_trade_ratio'
 'default_collar_fraction' 'dividend_yield_old' 'comp_float'
 'market_cap_old' 'pb_ratio_old' 'pe_ratio_old' 'shares_outstanding'
 'num_employees' 'year_founded']
ticker - 0%
name_long - 0%
name_short - 0%
country - 0%
margin_initial_ratio - 0%
maintenance_ratio - 0%
day_trade_ratio - 0%
list_date - 0%
investment_type - 0%
default_collar_fraction - 0%
dividend_yield_old - 40%
comp_float - 0%
market_cap_old - 0%
pb_ratio_old - 0%
pe_ratio_old - 23%
shares_outstanding - 0%
description - 0%
ceo - 0%
sector - 0%
industry - 0%
num_employees - 3%
year_founded - 2%


Based on the above analysis, dividend_yield_old has 40% of the data missing and pe_ratio_old has 23% of data missing. Losing high percentage of data is not recommended. I will also be adding new variables for most current values of these old variables through webpage resource later. There is also small percentage of data that is missing in num_employees and year_founded variables. I will be deleting those rows. 

In [13]:
# drop columns
df_missing = df_missing.drop(['dividend_yield_old', 'pe_ratio_old'], axis=1)
df_missing.shape

(1451, 20)

In [14]:
# dropping rows
df1 = df_missing.dropna(axis=0)
df1.shape

(1388, 20)

In [15]:
Datatype(df1)

There are 1388 rows and 20 columns
ticker                      object
name_long                   object
name_short                  object
country                     object
margin_initial_ratio       float64
maintenance_ratio          float64
day_trade_ratio            float64
list_date                   object
investment_type             object
default_collar_fraction    float64
comp_float                 float64
market_cap_old             float64
pb_ratio_old               float64
shares_outstanding         float64
description                 object
ceo                         object
sector                      object
industry                    object
num_employees              float64
year_founded               float64
dtype: object
['margin_initial_ratio' 'maintenance_ratio' 'day_trade_ratio'
 'default_collar_fraction' 'comp_float' 'market_cap_old' 'pb_ratio_old'
 'shares_outstanding' 'num_employees' 'year_founded']
['ticker' 'name_long' 'name_short' 'country' 'list_date' 'inves

### 6. Stripping string data of whitespace

In [16]:
def no_whitespace(string):
    return string.strip()

In [17]:
# stripping white space from all column type as object
def clean_string(df):
    
    for col in df.columns:
        if df[col].dtype == np.object:
            new = '{}_new'.format(col)
            df[new] = df[col].apply(no_whitespace)
            df[col] = df[new]
            df = df.drop(columns=[new])
    return df


df1 = clean_string(df1)
df1.head(2)

ticker                           name_long       name_short country  \
0    ACB  Aurora Cannabis Inc. Common Shares  Aurora Cannabis      CA   
1      F                  Ford Motor Company             Ford      US   

   margin_initial_ratio  maintenance_ratio  day_trade_ratio  list_date  \
0                   1.0               1.00             0.25  7/10/2014   
1                   0.5               0.25             0.25   1/2/1990   

  investment_type  default_collar_fraction    comp_float  market_cap_old  \
0           stock                     0.05  1.134714e+09    1.576224e+09   
1           stock                     0.05  3.887921e+09    2.759671e+10   

   pb_ratio_old  shares_outstanding  \
0      0.536001        1.168980e+09   
1      0.883266        3.965050e+09   

                                         description               ceo  \
0  Aurora Cannabis, Inc. engages in the productio...    Michael Singer   
1  Ford Motor Co. engages in the manufacture, dis...  James P. Hackett   

               sector                          industry  num_employees  \
0  Process Industries  Agricultural Commodities/Milling         2779.0   
1   Consumer Durables                    Motor Vehicles       190000.0   

   year_founded  
0        2006.0  
1        1903.0

### 7. Fuzzy Matching

Grouped by industry to find summary statistics

In [18]:
industry = df1.groupby('industry').describe()['market_cap_old']
industry

count          mean           std  \
industry                                                                      
Advertising/Marketing Services              3.0  7.497385e+09  5.903462e+09   
Aerospace & Defense                        23.0  2.421859e+10  3.796523e+10   
Aerospace And Defense                       6.0  2.022420e+10  3.674595e+10   
Agricultural Commodities Or Milling         3.0  4.769307e+09  2.627903e+09   
Agricultural Commodities/Milling            6.0  7.615452e+09  7.448627e+09   
...                                         ...           ...           ...   
Trucking                                    3.0  3.544373e+09  2.963817e+09   
Trucks Or Construction Or Farm Machinery    1.0  2.339878e+10           NaN   
Trucks/Construction/Farm Machinery         14.0  1.424684e+10  2.011440e+10   
Wholesale Distributors                     13.0  5.199139e+09  6.208347e+09   
Wireless Telecommunications                 7.0  2.160494e+10  2.802337e+10   

                                                   min           25%  \
industry                                                               
Advertising/Marketing Services            1.240085e+09  4.762003e+09   
Aerospace & Defense                       1.206279e+09  2.608919e+09   
Aerospace And Defense                     2.362140e+09  4.681463e+09   
Agricultural Commodities Or Milling       1.828824e+09  3.709745e+09   
Agricultural Commodities/Milling          1.438508e+09  2.235367e+09   
...                                                ...           ...   
Trucking                                  1.076555e+09  1.900698e+09   
Trucks Or Construction Or Farm Machinery  2.339878e+10  2.339878e+10   
Trucks/Construction/Farm Machinery        1.556107e+09  2.727970e+09   
Wholesale Distributors                    1.453217e+09  1.698788e+09   
Wireless Telecommunications               2.165936e+09  3.135855e+09   

                                                   50%           75%  \
industry                                                               
Advertising/Marketing Services            8.283921e+09  1.062604e+10   
Aerospace & Defense                       6.577388e+09  2.910462e+10   
Aerospace And Defense                     5.009404e+09  8.226157e+09   
Agricultural Commodities Or Milling       5.590666e+09  6.239548e+09   
Agricultural Commodities/Milling          5.384498e+09  9.817451e+09   
...                                                ...           ...   
Trucking                                  2.724841e+09  4.778281e+09   
Trucks Or Construction Or Farm Machinery  2.339878e+10  2.339878e+10   
Trucks/Construction/Farm Machinery        4.854832e+09  1.301976e+10   
Wholesale Distributors                    2.043538e+09  3.424355e+09   
Wireless Telecommunications               4.587412e+09  3.047422e+10   

                                                   max  
industry                                                
Advertising/Marketing Services            1.296815e+10  
Aerospace & Defense                       1.550000e+11  
Aerospace And Defense                     9.509370e+10  
Agricultural Commodities Or Milling       6.888430e+09  
Agricultural Commodities/Milling          2.100445e+10  
...                                                ...  
Trucking                                  6.831722e+09  
Trucks Or Construction Or Farm Machinery  2.339878e+10  
Trucks/Construction/Farm Machinery        6.834231e+10  
Wholesale Distributors                    1.965002e+10  
Wireless Telecommunications               7.726107e+10  

[152 rows x 8 columns]

In [19]:
for col in df1[['industry', 'sector']]:
        df1[col]
        print('Number of unique values in {}: {}'.format(str(col),df1[col].nunique()))

Number of unique values in industry: 152
Number of unique values in sector: 21


It shows data has 152 industries. However, there are some industry that seems to have duplicates with some mistakes, I will conduct fuzzy matching for this variable.

#### Token Sort Method

Part 1: Find 90% match using Token Sort method and replace values 

In [20]:
from fuzzywuzzy import process, fuzz

In [21]:
def token_high(df, percent):
    
    # create a subset dataframe with higher percent match values
    high_score = df[(df['score'] >= percent) &
                                  (df['industry'] !=  df['match']) &
                                  (df['sorted_industry'] != df['match'])]
    
    # Drop the representative value column
    high_score = high_score.drop('sorted_industry',axis=1).copy()
    
    return high_score

In [22]:
def token_fuzzy(df, percent, scorer):
    
    # unique industry lists
    unique_industry = df['industry'].unique().tolist()
    sorted(unique_industry)
    
    '''Token Ratio'''
    #Create tuples of industries, matched industries, and the score of token sort
    score = [(x,) + i
             for x in unique_industry 
             for i in process.extract(x, unique_industry, scorer=scorer)]

    #Create a dataframe from the tuples
    similarity = pd.DataFrame(score, columns=['industry','match','score'])

    # because this will have repitative pairs, we will remove duplicated pairs from this
    similarity['sorted_industry'] = np.minimum(similarity['industry'], similarity['match'])
    
    # Token high score
    high_score = token_high(similarity, percent)
    
    
    return similarity, high_score

In [23]:
similarity, high_score = token_fuzzy(df1, 90, fuzz.token_sort_ratio)
high_score.groupby(['industry','score']).agg({'match': ', '.join}).sort_values(['score'], ascending=False)

match
industry                                 score                                    
Chemicals Agricultural                   100               Chemicals: Agricultural
Pharmaceuticals Other                    100                Pharmaceuticals: Other
Auto Parts OEM                           100                       Auto Parts: OEM
Pharmaceuticals Major                    100                Pharmaceuticals: Major
Beverages Alcoholic                      100                  Beverages: Alcoholic
Chemicals Specialty                      100                  Chemicals: Specialty
Chemicals Major Diversified              100          Chemicals: Major Diversified
Agricultural Commodities Or Milling      96       Agricultural Commodities/Milling
Electronic Equipment Or Instruments      96       Electronic Equipment/Instruments
Property Or Casualty Insurance           95            Property/Casualty Insurance
Insurance Brokers Or Services            95             Insurance Brokers/Services
Internet Software Or Services            95             Internet Software/Services
Oilfield Services Or Equipment           95            Oilfield Services/Equipment
Investment Banks Or Brokers              94               Investment Banks/Brokers
Office Equipment Or Supplies             94              Office Equipment/Supplies
Apparel Or Footwear Retail               94                Apparel/Footwear Retail
Household Or Personal Care               94                Household/Personal Care
Containers Or Packaging                  93                   Containers/Packaging
Food Specialty Or Candy                  93                  Food: Specialty/Candy
Movies Or Entertainment                  93                   Movies/Entertainment
Trucks Or Construction Or Farm Machinery 92     Trucks/Construction/Farm Machinery
Cable Or Satellite TV                    92                     Cable/Satellite TV
Apparel Or Footwear                      91                       Apparel/Footwear
Oil & Gas Production                     90                 Oil And Gas Production
Beverages Alcoholic                      90               Beverages: Non-Alcoholic
Beverages: Alcoholic                     90               Beverages: Non-Alcoholic
Casinos Or Gaming                        90                         Casinos/Gaming

Replacing Values in original dataframe

In [24]:
# create dictionary
match_dict = dict(zip(high_score.industry, high_score.match))
len(match_dict)

26

In [25]:
# deleting to it can be added with a different name
del match_dict['Oil & Gas Production']

# deleting values that seems to be false positive
del match_dict['Beverages Alcoholic']
del match_dict['Beverages: Alcoholic']

In [26]:
len(match_dict)

23

In [27]:
match_dict['Oil And Gas Production'] = 'Oil & Gas Production'
match_dict['Beverages Alcoholic'] = 'Beverages: Alcoholic'

In [28]:
df1["industry"].replace(match_dict, inplace=True)

In [29]:
for col in df1[['industry', 'sector']]:
        df1[col]
        print('Number of unique values in {}: {}'.format(str(col),df1[col].nunique()))

Number of unique values in industry: 127
Number of unique values in sector: 21


Part 2: Find 80% match using Token Sort method and replace values 

In [30]:
similarity, high_score = token_fuzzy(df1, 80, fuzz.token_sort_ratio)
high_score.groupby(['industry','score']).agg({'match': ', '.join}).sort_values(['score'], ascending=False)

match
industry                     score                            
Beverages: Alcoholic         90       Beverages: Non-Alcoholic
Aerospace & Defense          89          Aerospace And Defense
Oil & Gas Pipelines          89          Oil And Gas Pipelines
Finance Or Rental Or Leasing 88         Finance/Rental/Leasing
Tools & Hardware             88             Tools And Hardware
Pharmaceuticals: Major       86         Pharmaceuticals: Other
Electronics/Appliance Stores 84         Electronics/Appliances
Pulp & Paper                 83                 Pulp And Paper
Apparel/Footwear             82        Apparel/Footwear Retail
Other Consumer Services      82     Other Consumer Specialties
Pharmaceuticals: Generic     82         Pharmaceuticals: Other
Financial Conglomerates      81       Industrial Conglomerates

In [31]:
# create dictionary
match_dict = dict(zip(high_score.match, high_score.industry))
len(match_dict)

11

In [32]:
# deleing values that are either false positive or needs to have different name
del match_dict['Beverages: Non-Alcoholic']
del match_dict['Pharmaceuticals: Other']
del match_dict['Finance/Rental/Leasing']
del match_dict['Electronics/Appliances']
del match_dict['Apparel/Footwear Retail']
del match_dict['Industrial Conglomerates']

In [33]:
match_dict['Finance Or Rental Or Leasing'] = 'Finance/Rental/Leasing'
match_dict['Electronics/Appliance Stores'] = 'Electronics/Appliances'
match_dict['Apparel/Footwear'] = 'Apparel/Footwear Retail'

In [34]:
df1["industry"].replace(match_dict, inplace=True)

In [35]:
for col in df1[['industry', 'sector']]:
        df1[col]
        print('Number of unique values in {}: {}'.format(str(col),df1[col].nunique()))

Number of unique values in industry: 119
Number of unique values in sector: 21


#### Part 2 Token Set Method

Find 90% match using Token Sort method and replace values 

In [36]:
similarity, high_score = token_fuzzy(df1, 90, fuzz.token_set_ratio)
high_score.groupby(['industry','score']).agg({'match': ', '.join}).sort_values(['score'], ascending=False)

match
industry                         score                             
Beverages: Alcoholic             100       Beverages: Non-Alcoholic
Hotels Or Resorts Or Cruiselines 93     Hotels/Resorts/Cruise lines

In [37]:
# replacing only "Hotels Or Resorts Or Cruiselines"
df1["industry"].replace({"Hotels Or Resorts Or Cruiselines": "Hotels/Resorts/Cruise lines"}, inplace=True)

2. Find 80% match using Token Sort method and replace values 

In [38]:
similarity, high_score = token_fuzzy(df1, 80, fuzz.token_set_ratio)
high_score.groupby(['industry','score']).agg({'match': ', '.join}).sort_values(['score'], ascending=False)

match
industry                     score                                                   
Beverages: Alcoholic         100                             Beverages: Non-Alcoholic
Chemicals: Major Diversified 87                               Food: Major Diversified
Electronics Distributors     86                                      Gas Distributors
Food Distributors            86                                      Gas Distributors
Gas Distributors             86          Wholesale Distributors, Medical Distributors
Major Telecommunications     86     Telecommunications Equipment, Wireless Telecom...
Pharmaceuticals: Major       86                                Pharmaceuticals: Other
Electronics Distributors     83                                     Food Distributors
Food Distributors            83          Wholesale Distributors, Medical Distributors
Pharmaceuticals: Generic     83        Pharmaceuticals: Major, Pharmaceuticals: Other
Financial Conglomerates      81         Media Conglomerates, Industrial Conglomerates
Industrial Conglomerates     81                                   Media Conglomerates
Specialty Telecommunications 80                           Wireless Telecommunications
Telecommunications Equipment 80                           Wireless Telecommunications

None of the values in 80% match needs to be deleted.

Grouped by sector to find summary statistics

In [39]:
sector = df1.groupby('sector').describe()['market_cap_old']
sector

count          mean           std           min  \
sector                                                                    
Commercial Services      29.0  7.429992e+09  1.198965e+10  1.240085e+09   
Communications           16.0  4.538270e+10  7.963304e+10  2.165936e+09   
Consumer Durables        53.0  1.108573e+10  1.910150e+10  1.089127e+09   
Consumer Non Durables     5.0  7.021542e+10  1.244877e+11  1.923602e+09   
Consumer Non-Durables    69.0  2.373423e+10  4.155025e+10  1.063352e+09   
Consumer Services        94.0  1.684320e+10  3.733050e+10  1.009149e+09   
Distribution Services    25.0  7.672228e+09  8.651178e+09  1.023215e+09   
Electronic Technology   137.0  2.594795e+10  1.063884e+11  1.006107e+09   
Energy Minerals          39.0  2.029769e+10  4.362654e+10  1.131418e+09   
Finance                 212.0  2.347638e+10  6.732858e+10  1.016566e+09   
Health Services          15.0  3.702691e+10  6.103079e+10  1.331306e+09   
Health Technology       142.0  2.211647e+10  4.657558e+10  1.008716e+09   
Industrial Services      36.0  1.257213e+10  1.667436e+10  1.006245e+09   
Non Energy Minerals       3.0  6.117305e+09  5.592933e+09  2.405740e+09   
Non-Energy Minerals      43.0  6.165261e+09  7.411135e+09  1.011189e+09   
Process Industries       63.0  8.786104e+09  1.161724e+10  1.030598e+09   
Producer Manufacturing  106.0  1.094236e+10  2.021695e+10  1.026670e+09   
Retail Trade             68.0  3.631207e+10  1.218593e+11  1.004924e+09   
Technology Services     160.0  3.969019e+10  1.471539e+11  1.037996e+09   
Transportation           15.0  2.233197e+10  3.143575e+10  1.062759e+09   
Utilities                58.0  1.693314e+10  2.179153e+10  1.402798e+09   

                                 25%           50%           75%           max  
sector                                                                          
Commercial Services     2.197509e+09  3.916458e+09  7.117903e+09  6.482886e+10  
Communications          3.240841e+09  1.073661e+10  3.826186e+10  2.600000e+11  
Consumer Durables       1.910831e+09  4.162557e+09  1.219081e+10  1.230000e+11  
Consumer Non Durables   6.738482e+09  1.992366e+10  3.049135e+10  2.920000e+11  
Consumer Non-Durables   2.477056e+09  6.019701e+09  2.841986e+10  2.290000e+11  
Consumer Services       2.148894e+09  4.914600e+09  1.292333e+10  2.120000e+11  
Distribution Services   1.982557e+09  3.424355e+09  1.268313e+10  3.389212e+10  
Electronic Technology   2.411008e+09  5.461357e+09  1.685482e+10  1.200000e+12  
Energy Minerals         2.293654e+09  6.218637e+09  1.671713e+10  2.180000e+11  
Finance                 1.969687e+09  3.791064e+09  1.660272e+10  5.020000e+11  
Health Services         4.780963e+09  1.415424e+10  4.260763e+10  2.420000e+11  
Health Technology       1.884791e+09  3.998896e+09  1.547975e+10  3.540000e+11  
Industrial Services     1.787846e+09  5.329576e+09  1.533769e+10  7.536015e+10  
Non Energy Minerals     2.900904e+09  3.396068e+09  7.973088e+09  1.255011e+10  
Non-Energy Minerals     1.580032e+09  2.877954e+09  6.582161e+09  3.604244e+10  
Process Industries      2.119767e+09  4.683380e+09  8.858107e+09  5.589628e+10  
Producer Manufacturing  2.027498e+09  3.752146e+09  8.012784e+09  1.160000e+11  
Retail Trade            2.745491e+09  5.629039e+09  1.959490e+10  9.380000e+11  
Technology Services     2.504907e+09  6.450685e+09  1.689845e+10  1.230000e+12  
Transportation          1.799007e+09  6.831722e+09  3.319485e+10  1.100000e+11  
Utilities               3.556573e+09  8.214063e+09  2.327308e+10  1.240000e+11

There are also two sectors that seems repetitive which I will replace it instead of doing fuzzy matching because it's only two values.

In [40]:
df1["sector"].replace({"Consumer Non Durables": "Consumer Non-Durables"}, inplace=True)
df1["sector"].replace({"Non Energy Minerals": "Non-Energy Minerals"}, inplace=True)

In [41]:
df1.head()

ticker                           name_long       name_short country  \
0    ACB  Aurora Cannabis Inc. Common Shares  Aurora Cannabis      CA   
1      F                  Ford Motor Company             Ford      US   
2     GE            General Electric Company               GE      US   
3   MSFT  Microsoft Corporation Common Stock        Microsoft      US   
5    FIT   Fitbit, Inc. Class A common stock           Fitbit      US   

   margin_initial_ratio  maintenance_ratio  day_trade_ratio  list_date  \
0                   1.0               1.00             0.25  7/10/2014   
1                   0.5               0.25             0.25   1/2/1990   
2                   0.5               0.25             0.25   1/2/1990   
3                   0.5               0.25             0.25  9/17/1987   
5                   0.5               0.25             0.25  6/18/2015   

  investment_type  default_collar_fraction    comp_float  market_cap_old  \
0           stock                     0.05  1.134714e+09    1.576224e+09   
1           stock                     0.05  3.887921e+09    2.759671e+10   
2           stock                     0.05  8.703145e+09    9.509370e+10   
3           stock                     0.05  7.489814e+09    1.230000e+12   
5           stock                     0.05  2.328025e+08    1.695109e+09   

   pb_ratio_old  shares_outstanding  \
0      0.536001        1.168980e+09   
1      0.883266        3.965050e+09   
2      3.460190        8.738434e+09   
3     11.943700        7.606050e+09   
5      3.528060        2.652752e+08   

                                         description  \
0  Aurora Cannabis, Inc. engages in the productio...   
1  Ford Motor Co. engages in the manufacture, dis...   
2  General Electric Co. is a technology and finan...   
3  Microsoft Corp. engages in the development and...   
5  Fitbit, Inc. engages in the development of wea...   

                          ceo                 sector  \
0              Michael Singer     Process Industries   
1            James P. Hackett      Consumer Durables   
2  H. Lawrence Culp, Jr., MBA  Electronic Technology   
3               Satya Nadella    Technology Services   
5                  James Park      Consumer Durables   

                           industry  num_employees  year_founded  
0  Agricultural Commodities/Milling         2779.0        2006.0  
1                    Motor Vehicles       190000.0        1903.0  
2               Aerospace & Defense       205000.0        1878.0  
3                 Packaged Software       144000.0        1975.0  
5           Other Consumer Services         1684.0        2007.0

In [42]:
df1.to_csv('clean_file.csv', encoding='utf-8', index=False)

### Summary of the Milestone 2

I have performed the following steps:
1. Reading the data file
2. Renaming data headers
3. Checked Duplicates 
4. Subset and filtered the dataset into smaller set with companies market cap over 1B
5. Handled missing values
6. Striped String data of whitespace and
7. Performed Fuzzy Matching for Industry columns.  
  
I have not looked for outliers because the data set mainly contains metadata and other non-current information. I have also changed the source for my flat file because after I cleaned the data, my original CSV file did not contained any important feature that can be combined with other sources to create graphs. I found similar dataset with additional features and decided to switch the data. Here is the source link for the dataset: [Robinhood data](https://www.kaggle.com/vaghefi/robinhood).
For Fuzzy Matching, I used some elements of following [GitHub](https://github.com/thuynh323/Natural-language-processing/blob/master/FuzzyWuzzy%20-%20Ramen%20Rater%20List/Find%20similar%20strings%20with%20FuzzyWuzzy.ipynb) tutorial